### ISSUE TITLE
CA1816 (Call GC.SuppressFinalize correctly) should not be reported when no finalizer is supported

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn-analyzers#4801]
### Analyzer

**Diagnostic ID**: [CA1816](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1816)

### Describe the improvement

CA1816 should only be reported in cases where users are following a design pattern intended to support the use of finalizers for releasing unmanaged resources. When the pattern is clearly not meant to support this case, the warning should be automatically suppressed.

### Describe suggestions on how to achieve the rule

A type meeting the following all of the following conditions should not report CA1816:

1. The type does not declare a user-defined finalizer
2. The type implements `IDisposable.Dispose()`, and this implementation is `sealed`

### Additional context

The second rule (detecting a `sealed` implementation of `Dispose()`) was intentionally chosen over looking for a `Dispose(bool)` method, since it more precisely identifies the API characteristic of `Dispose()` implementations that originally led to the guideline to always call `GC.SuppressFinalize()`.

**Comment by Youssef1313** [comment]
Is this a dupe of #3925?

**Comment by mavasani** [comment]
@stephentoub do you have any objection to this rule?

**Comment by stephentoub** [comment]
If a type or any of its base type hierarchy (that latter part isn't explicitly called out in Sam's description) doesn't define a finalizer, and the type is sealed, then it's not finalizable, so there's no need to warn about GC.SuppressFinalize in its Dispose implementation.  I'm ok with not warning in such a case.

cc: @bartonjs

**Comment by sharwell** [comment]
@stephentoub when the _type_ is sealed, we already do not warn. I made a mistake in my original proposal: the case that today warns but I propose should not warn is a non-sealed type with a virtual (non-sealed) implementation of `IDisposable.Dispose()`. The original post above has been updated to reflect the intended change.

**Comment by bartonjs** [comment]
Do we already have an analyzer that detects the non-canonical state of "you're adding a finalizer, your type hierarchy is disposable via virtual dispose instead of the basic dispose pattern, so figure out where to add GC.SuppressFinalize(this)"?  To me, that's a prerequisite.

Since we recommend the basic dispose pattern in all non-sealed cases of IDisposable I'm, personally, OK with extra warnings here.  But if we're going to make the analyzers nicer to non-canonical code we shouldn't regress the corner cases where the warnings were still useful.

**Comment by sharwell** [comment]
> Do we already have an analyzer that detects the non-canonical state of "you're adding a finalizer, your type hierarchy is disposable via virtual dispose instead of the basic dispose pattern, so figure out where to add GC.SuppressFinalize(this)"? 

[CA1063 (Implement IDisposable correctly)](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1063) reports a diagnostic if `Dispose()` is not sealed. In other words, the case I'm proposing to disable for CA1816 in cross-covered by another design rule.

With that said, note that CA1063 by default only looks at exposed types. That shouldn't be a problem; users intending to support non-canonical forms of `IDisposable` either internally or publicly would be likely to benefit from the change I proposed above.

**Comment by FlorianGrimm** [comment]
If a class implements IDisposable, but don't have a finalizer 
than there is no need for SuppressFinalize since it doesn't have any.
e.g. https://github.com/dotnet/reactive/blob/main/Rx.NET/Source/src/System.Reactive/Subjects/SubjectBase.cs
Which also often apply to any unit test fakes.

**Comment by sharwell** [comment]
> If a class implements IDisposable, but don't have a finalizer than there is no need for SuppressFinalize since it doesn't have any.

This is not correct for types implementing the [dispose pattern](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern), and is covered by the [CA1816](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1816) documentation.

**Comment by bartonjs** [comment]
Since `SubjectBase` isn't `sealed`, a derived type could add a finalizer.  Since the call to SuppressFinalize should be in the non-virtual Dispose that can't be left as a detail to the type that added the finalizer, and is why the pattern is what it is.

**Comment by sharwell** [comment]
@bartonjs Note that SubjectBase does not implement the dispose pattern. It provides a virtual `Dispose()` method, a non-standard pattern which disallows the introduction of a finalizer in a derived type.
